repository/src/main/java/org/apache/atlas/repository/store/graph/v2/preprocessor/datamesh/DataProductPreProcessor.java

```diff
@@ -0,0 +1,178 @@
+
+            if(relationshipAttribute instanceof AtlasObjectId){
+                AtlasObjectId objectId = (AtlasObjectId) relationshipAttribute;
+                if (objectId != null) {
+                    if (StringUtils.isNotEmpty(objectId.getGuid())) {
+                        AtlasVertex vertex = entityRetriever.getEntityVertex(objectId.getGuid());
+
+                        if (vertex == null) {
+                            parentDomain = entityRetriever.toAtlasEntityHeader(objectId.getGuid(), attributes);
+                        } else {
+                            parentDomain = entityRetriever.toAtlasEntityHeader(vertex, attributes);
+                        }
+
+                    } else if (MapUtils.isNotEmpty(objectId.getUniqueAttributes()) &&
+                            StringUtils.isNotEmpty((String) objectId.getUniqueAttributes().get(QUALIFIED_NAME))) {
+                        AtlasVertex parentDomainVertex = entityRetriever.getEntityVertex(objectId);
+                        parentDomain = entityRetriever.toAtlasEntityHeader(parentDomainVertex, attributes);
+
+                    }
+                }
+            }
+            else if(relationshipAttribute instanceof Map){
+                Map<String, Object> relationshipMap = (Map<String, Object>) relationshipAttribute;
+                if (StringUtils.isNotEmpty((String) relationshipMap.get("guid"))) {
+                    AtlasVertex vertex = entityRetriever.getEntityVertex((String) relationshipMap.get("guid"));
+
+                    if (vertex == null) {
+                        parentDomain = entityRetriever.toAtlasEntityHeader((String) relationshipMap.get("guid"), attributes);
+                    } else {
+                        parentDomain = entityRetriever.toAtlasEntityHeader(vertex, attributes);
+                    }
+
+                }
+                else  {
+                    parentDomain = new AtlasEntityHeader((String) relationshipMap.get("typeName"), relationshipMap);
+
+                }
+            }
+            else{
+                LOG.warn("DataProductPreProcessor.setParent: Invalid relationshipAttribute {}", relationshipAttribute);
+            }
+
+        }
+        RequestContext.get().endMetricRecord(metricRecorder);
+    }
+
+    private void productExists(String productName, String parentDomainQualifiedName) throws AtlasBaseException {
+        AtlasPerfMetrics.MetricRecorder metricRecorder = RequestContext.get().startMetricRecord("domainExists");
+
+        boolean exists = false;
+        try {
+            List mustClauseList = new ArrayList();
+            mustClauseList.add(mapOf("term", mapOf("__typeName.keyword", DATA_PRODUCT_ENTITY_TYPE)));
+            mustClauseList.add(mapOf("term", mapOf("__state", "ACTIVE")));
+            mustClauseList.add(mapOf("term", mapOf("name.keyword", productName)));
+
+
+            Map<String, Object> bool = new HashMap<>();
+            if (parentDomain != null) {
```

Comment:
Also, parentDomainQualifiedName should never be empty, Assume we will always get parentDomainQualifiedName here, if not  throw an exception here

Reply:
In case of root domain we wont get parentDomainQualifiedName